scIB/metrics/trajectory.py

```diff
@@ -67,37 +68,61 @@ def trajectory_conservation(
-
-    sc.tl.dpt(adata_post_sub2)  # stored in 'dpt_pseudotime'
-    adata_post_sub2.obs.loc[adata_post_sub2.obs['dpt_pseudotime'] > 1, 'dpt_pseudotime'] = 0
-    adata_post_sub.obs['dpt_pseudotime'] = 0
-    adata_post_sub.obs['dpt_pseudotime'] = adata_post_sub2.obs['dpt_pseudotime']
-    adata_post_sub.obs['dpt_pseudotime'].fillna(0, inplace=True)
-
-    pseudotime_before = adata_post_sub.obs['dpt_pseudotime']
-    pseudotime_after = adata_pre_sub.obs[pseudotime_key]
-    correlation = pseudotime_before.corr(pseudotime_after, 'spearman')
-
-    return (correlation + 1) / 2  # scaled
-
+    if batch_key==False:
+        try:
+            iroot, adata_post_sub2 = get_root(adata_pre_sub, adata_post_sub, label_key, pseudotime_key)
+        except RootCellError:
+            print('No root cell found, setting trajectory conservation metric to 0.')
+            return 0  # failure to find root cell means no TI conservation
+
+        adata_post_sub2.uns['iroot'] = iroot
+
+        sc.tl.dpt(adata_post_sub2)  # stored in 'dpt_pseudotime'
+        adata_post_sub2.obs.loc[adata_post_sub2.obs['dpt_pseudotime'] > 1, 'dpt_pseudotime'] = 0
+        adata_post_sub.obs['dpt_pseudotime'] = 0
+        adata_post_sub.obs['dpt_pseudotime'] = adata_post_sub2.obs['dpt_pseudotime']
+        adata_post_sub.obs['dpt_pseudotime'].fillna(0, inplace=True)
+
+        pseudotime_before = adata_post_sub.obs['dpt_pseudotime']
+        pseudotime_after = adata_pre_sub.obs[pseudotime_key]
+        correlation = pseudotime_before.corr(pseudotime_after, 'spearman')
+
+        return (correlation + 1) / 2  # scaled
+    else:
+        corr = pd.Series()
+        for i in batch_key:
+            adata_pre_sub_tmp = adata_pre_sub[adata_pre_sub.obs[batch_key]==i]
+            adata_post_sub_tmp = adata_post_sub[adata_post_sub.obs[batch_key]==i]
+            try:
+                iroot, adata_post_sub2_tmp = get_root(adata_pre_sub_tmp, adata_post_sub_tmp, label_key, pseudotime_key)
+            except RootCellError:
+                print('No root cell found, setting trajectory conservation metric to 0.')
+                return 0  # failure to find root cell means no TI conservation
+            
+            adata_post_sub2_tmp.uns['iroot'] = iroot
+            
+            sc.tl.dpt(adata_post_sub2_tmp)  # stored in 'dpt_pseudotime'
+            adata_post_sub2_tmp.obs.loc[adata_post_sub2_tmp.obs['dpt_pseudotime'] > 1, 'dpt_pseudotime'] = 0
+            adata_post_sub_tmp.obs['dpt_pseudotime'] = 0
+            adata_post_sub_tmp.obs['dpt_pseudotime'] = adata_post_sub2_tmp.obs['dpt_pseudotime']
+            adata_post_sub_tmp.obs['dpt_pseudotime'].fillna(0, inplace=True)
+            
+            pseudotime_before = adata_post_sub_tmp.obs['dpt_pseudotime']
+            pseudotime_after = adata_pre_sub_tmp.obs[pseudotime_key]
+            corr[i] = pseudotime_before.corr(pseudotime_after, 'spearman')
+        return (corr.mean() + 1) / 2 # scaled
```

Comment:
I'm assuming this part of the code is a duplicate of the code above, just called per batch. It would be great if you could put this into a separate function and call that function twice instead. That way it should be easier to read